File: examples/python/commitments_example.py
import pulumi
import pulumi_castai as castai

# Placeholder: Assume cluster_id is obtained from an existing cluster resource
# e.g., eks_cluster = castai.aws.EksCluster(...)
# cluster_id = eks_cluster.id
cluster_id = "your-cluster-id" # Replace with a valid cluster ID

# Create a commitment plan
commitments_kwargs = {
    "name": "production-py-commitments",
    "auto_apply": True,
    "auto_apply_label": "environment=production",
    "compute_unit_hourly_rate": 0.0294, # Example rate, replace with actual value
    "instances": [{
        "cluster_id": cluster_id,
        "compute_units": 100,
        "start_date": "2024-01-01T00:00:00Z", # Adjust dates as needed
        "end_date": "2024-12-31T23:59:59Z",
    }]
}

commitments = castai.Commitments("example-py-commitments", **commitments_kwargs)

# Create a reservation of compute instances
reservations_kwargs = {
    "cluster_id": cluster_id,
    "instances": [{
        "instance_type": "m5.2xlarge",
        "provider": "aws", # Or gcp, azure
        "zone": "us-west-2a", # Adjust zone
        "spot_instance": False,
        "count": 3,
        "start_date": "2024-01-01T00:00:00Z", # Adjust dates
        "end_date": "2024-06-30T23:59:59Z",
        "labels": {
            "reserved": "true",
            "workload": "database"
        },
        "taints": [{
            "key": "reserved",
            "value": "database",
            "effect": "NoSchedule"
        }]
    }],
    # "node_configuration_id": "your-node-config-id" # Optional
}

reservations = castai.Reservations("example-py-reservations", **reservations_kwargs)

# Export commitment and reservation IDs
pulumi.export("commitment_py_id", commitments.id)
pulumi.export("reservation_py_id", reservations.id) 

File: examples/go/commitments_example.go
package main

import (
	"github.com/castai/pulumi-castai/sdk/go/castai" // Assuming main package
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

// runCommitmentsExample demonstrates creating commitments and reservations.
// Renamed from main to avoid conflicts.
func runCommitmentsExample(ctx *pulumi.Context) error {
	// Placeholder: Assume clusterId is obtained from an existing cluster resource
	// e.g., eksCluster, err := aws.NewEksCluster(ctx, ...)
	// clusterId := eksCluster.ID()
	clusterId := pulumi.String("your-cluster-id") // Replace with a valid cluster ID

	// Create a commitment plan
	commitments, err := castai.NewCommitments(ctx, "example-go-commitments", &castai.CommitmentsArgs{
		Name:                  pulumi.String("production-go-commitments"),
		AutoApply:             pulumi.Bool(true),
		AutoApplyLabel:        pulumi.String("environment=production"),
		ComputeUnitHourlyRate: pulumi.Float64(0.0294), // Example rate, replace with actual value
		Instances: castai.CommitmentsInstanceArray{
			&castai.CommitmentsInstanceArgs{
				ClusterId:    clusterId,
				ComputeUnits: pulumi.Int(100),
				StartDate:    pulumi.String("2024-01-01T00:00:00Z"), // Adjust dates as needed
				EndDate:      pulumi.String("2024-12-31T23:59:59Z"),
			},
		},
	})
	if err != nil {
		return err
	}

	// Create a reservation of compute instances
	reservations, err := castai.NewReservations(ctx, "example-go-reservations", &castai.ReservationsArgs{
		ClusterId: clusterId,
		Instances: castai.ReservationsInstanceArray{
			&castai.ReservationsInstanceArgs{
				InstanceType: pulumi.String("m5.2xlarge"),
				Provider:     pulumi.String("aws"),        // Or gcp, azure
				Zone:         pulumi.String("us-west-2a"), // Adjust zone
				SpotInstance: pulumi.Bool(false),
				Count:        pulumi.Int(3),
				StartDate:    pulumi.String("2024-01-01T00:00:00Z"), // Adjust dates
				EndDate:      pulumi.String("2024-06-30T23:59:59Z"),
				Labels: pulumi.StringMap{
					"reserved": pulumi.String("true"),
					"workload": pulumi.String("database"),
				},
				Taints: castai.ReservationsInstanceTaintArray{
					&castai.ReservationsInstanceTaintArgs{
						Key:    pulumi.String("reserved"),
						Value:  pulumi.String("database"),
						Effect: pulumi.String("NoSchedule"),
					},
				},
			},
		},
		// NodeConfigurationId: pulumi.String("your-node-config-id"), // Optional
	})
	if err != nil {
		return err
	}

	// Export commitment and reservation IDs
	ctx.Export("commitment_go_id", commitments.ID())
	ctx.Export("reservation_go_id", reservations.ID())

	return nil
}

/*
// Example of how you might call this from your main function:
func main() {
    pulumi.Run(func(ctx *pulumi.Context) error {
        if err := runCommitmentsExample(ctx); err != nil {
            return err
        }
        // ... potentially call other example functions ...
        return nil
    })
}
*/
